google/auth/transport/_custom_tls_signer.py

```diff
@@ -0,0 +1,234 @@
+# `SignForPython` method from the signer lib.
+def get_sign_callback(signer_lib, config_file_path):
+    def sign_callback(sig, sig_len, tbs, tbs_len):
+        _LOGGER.debug("calling sign callback...")
+
+        digest = _compute_sha256_digest(tbs, tbs_len)
+        digestArray = ctypes.c_char * len(digest)
+
+        # reserve 2000 bytes for the signature, shoud be more then enough.
+        # RSA signature is 256 bytes, EC signature is 70~72.
+        sigHolder = ctypes.create_string_buffer(2000)
+
+        sigLen = signer_lib.SignForPython(
+            config_file_path.encode(),  # configFilePath
+            digestArray.from_buffer(bytearray(digest)),  # digest
+            len(digest),  # digestLen
+            sigHolder,  # sigHolder
+            2000,  # sigHolderLen
+        )
+
+        if sigLen == 0:
+            # signing failed, return 0
+            return 0
+
+        sig_len[0] = sigLen
+        bs = bytearray(sigHolder)
+        for i in range(sigLen):
+            sig[i] = bs[i]
+
+        return 1
+
+    return SIGN_CALLBACK_CTYPE(sign_callback)
+
+
+# Obtain the certificate bytes by calling the `GetCertPemForPython` method from
+# the signer lib. The method is called twice, the first time is to compute the
+# cert length, then we create a buffer to hold the cert, and call it again to
+# fill the buffer.
+def get_cert(signer_lib, config_file_path):
+    # First call to calculate the cert length
+    certLen = signer_lib.GetCertPemForPython(
+        config_file_path.encode(),  # configFilePath
+        None,  # certHolder
+        0,  # certHolderLen
+    )
+    if certLen == 0:
+        raise exceptions.MutualTLSChannelError("failed to get certificate")
+
+    # Then we create an array to hold the cert, and call again to fill the cert
+    certHolder = ctypes.create_string_buffer(certLen)
+    signer_lib.GetCertPemForPython(
+        config_file_path.encode(),  # configFilePath
+        certHolder,  # certHolder
+        certLen,  # certHolderLen
+    )
+    return bytes(certHolder)
+
+
+class CustomTlsSigner(object):
+    def __init__(self, enterprise_cert_file_path):
+        """
+        This class loads the offload and signer library, and calls APIs from
+        these libraries to obtain the cert and a signing callback, and attach
+        them to SSL context. The cert and the signing callback will be used
+        for client authentication in TLS handshake.
+
+        Args:
+            enterprise_cert_file_path (str): the path to a enterprise cert JSON
+                file. The file should contain the following field:
+
+                    {
+                        "libs": {
+                            "signer_library": "...",
+                            "offload_library": "..."
+                        }
+                    }
+        """
+        self._enterprise_cert_file_path = enterprise_cert_file_path
+        self._cert = None
+        self._sign_callback = None
+
+    def load_libraries(self):
```

Comment:
why not load libraries in init?

Reply:
init is not easy to mock in general, since this is an internal module, we make it simple for testing purposes. We create a CustomTlsSigner object and call `load_libraries`, `set_up_custom_key` functions in `_MutualTlsOffloadAdapter`'s init instead.